README.md
# netty-ring-adapter

netty-ring-adapter is a ring server built with [Netty](https://netty.io/). netty-ring-adapter is designed to be a drop in ring
adapter that should work just like reference ring adapter.

[![Build Status](https://secure.travis-ci.org/RallySoftware/netty-ring-adapter.png)](http://travis-ci.org/RallySoftware/netty-ring-adapter)

## Installation

`netty-ring-adapter` is available as a Maven artifact from
[Clojars](http://clojars.org/netty-ring-adapter):

```clojure
[netty-ring-adapter "0.4.6"]
```

Previous versions available as

```clojure
[netty-ring-adapter "0.4.5"]
[netty-ring-adapter "0.4.4"]
[netty-ring-adapter "0.4.3"]
[netty-ring-adapter "0.4.2"]
```

## Usage

```clj
(use 'netty.ring.adapter)

(defn handler [req]
  {:status 200
   :headers {"Content-Type" "text/html"}
   :body "Hello world from Netty"})

(def shutdown (start-server handler {:port 8080}))

;; If you want to stop the server, just invoke the function returned from the `start-server` function.
(shutdown)
```

The server currently supports the following options when starting the server.

```clj
{ :port 8080                        ;; The port in which the server will be listening for requests
  :zero-copy true                   ;; Should the server send file response bodies with Netty's FileRegion functionality
  :channel-options                  ;; Channel options passed to the ServerBootstrap.setOptions
    { "child.tcpNoDelay" true}
  :max-http-chunk-length 1048576    ;; The maximum length of the aggregated content
  :number-of-handler-threads 16     ;; The number of threads that will be used to handle requests.
                                    ;; These threads are used to allow the handler function to work without blocking an I/O
                                    ;; worker thread.
  :max-channel-memory-size 1048576  ;; the maximum total size of the queued events per channel
  :max-total-memory-size 1048576    ;; the maximum total size of the queued events
  :debug :slf4j }                   ;; turns on debugging using the slf4j as a logging framework.
                                    ;; debugging options include (:commons :jboss :log4j :slf4j :jdk)
```

Using `:zero-copy` may not work in all cases depending on your operating system and JVM version. Please see
[FileRegion](http://static.netty.io/3.5/api/org/jboss/netty/channel/FileRegion.html) for more information.

## Development

To run the tests:

    $ lein deps
    $ lein test

## TODO

* ~~Support HTTP Keep-Alive? Not sure if this already works, but it should.~~
* Add metrics. Seems like it would be good to add JMX metrics to the various queues and requests/responses.
* Create a lein plugin for the netty-ring-adapter.
* SSL support


## License

Distributed under the Eclipse Public License, the same as Clojure. <http://opensource.org/licenses/eclipse-1.0.php>


HISTORY.md
## 0.4.6 (2013-04-23)
* Added a default backlog setting of 200

## 0.4.5 (2013-04-12)
* Updated Netty to 3.6.5-Final

## 0.4.4 (2013-03-20)
* Fixed a bug with file content type detection
* No longer trying to detect file content types.

## 0.4.3 (2013-03-09)
* Updated Netty to 3.6.3-Final
* Updated org.clojure/tools.logging to 0.2.6
* Started testing with the final release of Clojure 1.5

## 0.4.2 (2013-01-30)
* Fixed a defect that caused the server to consume CPU cycles when a client closes the socket before a response is sent.

## 0.4.1 (2013-01-28)
* Added error logging when exceptions are thrown during processing.

## 0.4.0 (2013-01-27)
* Added support for HTTP keep-alive

## 0.3.2 (2013-01-25)
* Added the ability to turn on debug logging (Netty logging).
* Debugging frameworks supported (:commons :jboss :log4j :slf4j :jdk)

## 0.3.0 (2013-01-24)
* Updated Netty to 3.6.2.Final
* All handler processing is now done in a different thread from the IO worker threads.
* Added many more options that can help configure the behavior of the adapter

```clj
{ :port 8080                        ;; The port in which the server will be listening for requests
  :zero-copy true                   ;; Should the server send file response bodies with Netty's FileRegion functionality
  :channel-options                  ;; Channel options passed to the ServerBootstrap.setOptions
    { "child.tcpNoDelay" true}
  :max-http-chunk-length 1048576    ;; The maximum length of the aggregated content
  :number-of-handler-threads 16     ;; The number of threads that will be used to handle requests.
                                    ;; These threads are used to allow the handler function to work without blocking an I/O
                                    ;; worker thread.
  :max-channel-memory-size 1048576  ;; the maximum total size of the queued events per channel
  :max-total-memory-size 1048576    ;; the maximum total size of the queued events
```

## 0.2.5 (2013-01-02)
* Updated Netty to 3.6.0.Final

## 0.2.4 (2012-10-29)
* Updated Netty to 3.5.9.Final

## 0.2.3 (2012-09-18)
* Updated license name

## 0.2.2 (2012-09-12)
* Updated Netty to 3.5.7.Final

## 0.2.1 (2012-08-25)

* Updated Netty to 3.5.5.Final
* Removed some reflection warnings
* Fixed bug that would cause the server to not respond if the response :body was null

## 0.2.0 (2012-08-22)

* netty-ring-adapter now supporting all the required [Ring SPEC](https://github.com/ring-clojure/ring/blob/master/SPEC) body types

## 0.1.0 (2012-08-19)

* First numbered release
* Currently only supports string response :body type


test/netty/ring/buffer_test.clj
(ns netty.ring.buffer-test
  (:use clojure.test)
  (:require [netty.ring.buffer :as buffer])
  (:import [org.jboss.netty.buffer ChannelBuffers ChannelBufferOutputStream]))

(declare to-channel-buffer write)

(deftest in-memory
  (let [in-memory-buffer (buffer/create-buffer 100)]
    (write in-memory-buffer "Hello")

    (is (= 100 (:max-in-memory-size @in-memory-buffer)))
    (is (= 5 (.readableBytes (:buffer @in-memory-buffer))))
    (is (= "Hello" (slurp (buffer/input-stream in-memory-buffer))))))

(deftest multiple-in-memory-writes
  (let [in-memory-buffer (buffer/create-buffer 1000)]
    (write in-memory-buffer "Hello ")
    (write in-memory-buffer "World")

    (is (= 11 (.readableBytes (:buffer @in-memory-buffer))))
    (is (= "Hello World" (slurp (buffer/input-stream in-memory-buffer))))))

(deftest file-backed
  (let [file-buffer (buffer/create-buffer 10)]
    (write file-buffer "I hope this works!")

    (is (= true (.isFile (:buffer @file-buffer))))
    (is (= "I hope this works!" (slurp (buffer/input-stream file-buffer))))))

(deftest multiple-writes-to-memory-and-file
  (let [file-buffer (buffer/create-buffer 30)]
    (write file-buffer "Hello ")
    (write file-buffer "World ")
    (write file-buffer "I should be writing to a file")

    (is (= true (.isFile (:buffer @file-buffer))))
    (is (= "Hello World I should be writing to a file" (slurp (buffer/input-stream file-buffer))))))

(defn write [buffer msg]
  (buffer/write buffer (to-channel-buffer msg)))

(defn to-channel-buffer [msg]
  (let [buffer (ChannelBuffers/dynamicBuffer)]
    (with-open [input-stream (ChannelBufferOutputStream. buffer)]
      (spit input-stream msg)
      buffer)))

test/netty/ring/adapter_test.clj
(ns netty.ring.adapter-test
  (:refer-clojure :exclude [get])
  (:use clojure.test
        netty.ring.adapter
        compojure.core)
  (:require [clj-http.client :as client]
            [compojure.route :as route]
            [clojure.java.io :as io]
            [ring.util.response :as response]))

(def ^:const server "http://localhost:8080")
(declare get post put make-request)

(deftest simple
  (is (= (get "/") "Hello World")))

(deftest request-method
  (is (= (get "/method") "get"))
  (is (= (post "/method") "post"))
  (is (= (put "/method") "put")))

(deftest uri
  (is (= (get "/uri/me") "/uri/me"))
  (is (= (get "/uri/you?help=me") "/uri/you")))

(deftest query-string
  (is (= (get "/query?you=me") "you=me"))
  (is (= (get "/query?me=you&you=I") "me=you&you=I")))

(deftest server-name
  (is (= (get "/serverName") "localhost")))

(deftest server-port
  (is (= (get "/port") "8080")))

(deftest remote-address
  (is (= (get "/remoteAddress") "127.0.0.1")))

(deftest scheme
  (is (= (get "/scheme") "http"))
  (is (= (make-request :get "/scheme" {:headers {"X-Scheme" "https"}}) "https")))

(deftest content-type
  (is (= (make-request :get "/contentType" {:content-type :json}) "application/json"))
  (is (= (make-request :get "/contentType" {:content-type :html}) "application/html")))

(deftest chracter-encoding
  (is (= (:body (client/post (str server "/characterEncoding") {:headers {"Content-Encoding" "UTF-8"}}) "UTF-8"))))

(deftest headers
  (is (= (get "/headers") "localhost:8080")))

(deftest request-body
  (is (= "Foo Bar" (make-request :post "/requestbody" {:body "Foo Bar"})))
  (is (= (slurp "./test/netty/ring/response.txt") (make-request :post "/requestbody" {:body (io/input-stream "./test/netty/ring/response.txt") :length -1}))))

(deftest response-headers
  (is (= "bar" (get-in (client/get (str server "/responseHeaders/single")) [:headers "foo"])))
  (is (= ["bar" "baz"] (get-in (client/get (str server "/responseHeaders/multiple")) [:headers "foo"]))))

(deftest response-body-types
  (is (= "agoodresponse" (get "/ISeqResponse")))
  (is (= "afineresponse" (get "/InputStreamResponse")))
  (is (= (slurp "./test/netty/ring/response.txt") (get "/FileResponse/response.txt")))
  (is (= (slurp "./test/netty/ring/response.json") (get "/FileResponse/response.json"))))

(deftest bad-responses
  (is (= "" (get "/EmptyResponse")))
  (is (= 500 (:status (client/get (str server "/Exception") {:throw-exceptions false})))))

(deftest keep-alive
  (client/with-connection-pool {:timeout 5 :threads 4 :insecure? false :default-per-route 10}
    (is (= "agoodresponse" (get "/ISeqResponse")))
    (is (= "keep-alive" (get-in (client/get (str server "/headers")) [:headers "connection"])))
    (is (= "afineresponse" (get "/InputStreamResponse")))
    (is (= "" (get "/EmptyResponse")))
    (is (= (slurp "./test/netty/ring/response.txt") (get "/FileResponse/response.txt")))))

(defn header-handler [request]
  (if (.contains (:uri request) "single")
    {:status 200 :headers {"foo" "bar"}}
    {:status 200 :headers {"foo" ["bar" "baz"]}}))

(defn exception-handler [request]
  (throw (Exception. "Bad things happen")))

(defroutes test-routes
  (GET "/" [] "Hello World")
  (ANY "/method" [] #(name (:request-method %)))
  (GET "/uri/*" [] #(:uri %))
  (GET "/query" [] #(:query-string %))
  (GET "/serverName" [] #(:server-name %))
  (GET "/port" [] #(str (:server-port %)))
  (GET "/remoteAddress" [] #(:remote-addr %))
  (GET "/scheme" [] #(name (:scheme %)))
  (GET "/contentType" [] #(:content-type %))
  (POST "/characterEncoding" [] #(:character-encoding %))
  (GET "/headers" [] #((:headers %) "host"))
  (POST "/requestbody" [] #(slurp (:body %)))
  (GET "/responseHeaders/*" [] header-handler)
  (GET "/ISeqResponse" [] {:status 200 :body '("a" "good" "response")})
  (GET "/InputStreamResponse" [] {:status 200 :body (io/input-stream (.getBytes "afineresponse"))})
  (GET "/FileResponse/:current-file" [current-file] (response/file-response (str "./test/netty/ring/" current-file)))
  (GET "/EmptyResponse" [] {:status 200 :body nil})
  (GET "/Exception" [] exception-handler)
  (route/not-found "Unknown"))

(defn server-fixture [f]
  (let [shutdown (start-server test-routes {:port 8080})]
    (f)
    (shutdown)))

(use-fixtures :each server-fixture)

(defn make-request [method path options]
  (:body (client/request (merge {:method method :url (str server path)} options))))

(defn request [f]
  #(:body (f (str server %))))

(def get (request client/get))
(def post (request client/post))
(def put (request client/put))


test/netty/ring/buffers_test.clj
(ns netty.ring.buffers-test
  (:require [netty.ring.buffers :as b])
  (:use clojure.test))

(deftest string-to-buffer
  (let [s "this is a string"
        buffer (b/to-buffer s)]
    (is (= s (b/buffer->string buffer)))))

(deftest seq-to-buffer
  (let [s '("this" "is" "a" "string")
        buffer (b/to-buffer s)]
    (is (= (apply str s) (b/buffer->string buffer)))))




src/netty/ring/buffer.clj
(ns netty.ring.buffer
  (:require [clojure.java.io :as io])
  (:import [org.jboss.netty.buffer ChannelBuffer ChannelBuffers ChannelBufferInputStream]
           [java.io File]))

(defprotocol Buffer
  (fill [buffer ^ChannelBuffer channel-buffer])
  (stream [buffer]))

(defn- create-file-buffer [^ChannelBuffer buffer ^ChannelBuffer channel-buffer]
  (let [file (File/createTempFile "netty" "buffer")
        buffers (into-array ChannelBuffer [buffer channel-buffer])
        composite-buffer-channel (ChannelBuffers/wrappedBuffer buffers)]
    (.deleteOnExit file)
    (with-open [buffer-stream (ChannelBufferInputStream. composite-buffer-channel)]
      (io/copy buffer-stream file))
    file))

(defn- create-in-memory-buffer [^Integer max-in-memory-size]
  (ChannelBuffers/dynamicBuffer max-in-memory-size))

(defn- full? [^ChannelBuffer buffer ^ChannelBuffer channel-buffer]
  (let [capacity (.capacity buffer)
        current-size (.readableBytes buffer)
        buffer-size (.readableBytes channel-buffer)]
    (> (+ buffer-size current-size) capacity)))

(extend-type ChannelBuffer
  Buffer
    (fill [buffer ^ChannelBuffer channel-buffer]
      (if-not (full? buffer channel-buffer)
        (doto buffer (.writeBytes channel-buffer))
        (create-file-buffer buffer channel-buffer)))
    (stream [buffer]
      (ChannelBufferInputStream. buffer)))

(extend-type File
  Buffer
    (fill [file-buffer ^ChannelBuffer channel-buffer]
      (with-open [output-stream (io/output-stream file-buffer)]
        (.getBytes channel-buffer 0 output-stream (.readableBytes channel-buffer)))
      file-buffer)
    (stream [file-buffer]
      (io/input-stream file-buffer)))

(defn create-buffer [max-in-memory-size]
  (atom {:max-in-memory-size max-in-memory-size
         :buffer nil}))

(defn write [buffer channel-buffer]
  (-> (or (:buffer @buffer) (create-in-memory-buffer (:max-in-memory-size @buffer)))
      (fill channel-buffer)
      (->> (swap! buffer assoc :buffer)))
  buffer)

(defn input-stream [buffer]
  (stream (:buffer @buffer)))

src/netty/ring/request.clj
(ns netty.ring.request
  (:require [clojure.string :as s])
  (:import [org.jboss.netty.buffer ChannelBufferInputStream]
           [org.jboss.netty.channel ChannelHandlerContext]
           [org.jboss.netty.handler.codec.http HttpMethod HttpRequest HttpHeaders HttpHeaders$Names]))

(def method-mapping
  {HttpMethod/GET :get
   HttpMethod/POST :post
   HttpMethod/PUT :put
   HttpMethod/TRACE :trace
   HttpMethod/PATCH :patch
   HttpMethod/OPTIONS :options
   HttpMethod/DELETE :delete
   HttpMethod/HEAD :head
   HttpMethod/CONNECT :connect})

(defn method [^HttpMethod method]
  (if-let [method-keyword (method-mapping method)]
    method-keyword
    (-> method (.getName) (s/lower-case) (keyword))))

(defn url [request-uri]
  (let [regex #"([^?]+)[?]?([^?]+)?"
        [match uri query] (re-find regex request-uri)]
    [uri query]))

(defn hostname [^HttpRequest request]
  (when-let [host (HttpHeaders/getHost request)]
    (get (s/split host #":") 0)))

(defn local-address [^ChannelHandlerContext context]
  (-> context .getChannel .getLocalAddress))

(defn server-name [^ChannelHandlerContext context ^HttpRequest request]
  (if-let [host (hostname request)]
    host
    (.getHostName (local-address context))))

(defn remote-address [^ChannelHandlerContext context]
  (-> context
    .getChannel
    .getRemoteAddress
    .getAddress
    .getHostAddress))

(defn scheme [^HttpRequest request]
  (let [scheme (HttpHeaders/getHeader request "X-Scheme" "http")]
    (keyword scheme)))

(defn content-type [^HttpRequest request]
  (if-let [type (HttpHeaders/getHeader request HttpHeaders$Names/CONTENT_TYPE)]
    (-> type
      (s/split #";")
      (get 0)
      s/trim
      s/lower-case)))

(defn content-length [^HttpRequest request]
  (let [length (HttpHeaders/getContentLength request)]
    (when (> length 0) length)))

(defn character-encoding [^HttpRequest request]
  (HttpHeaders/getHeader request HttpHeaders$Names/CONTENT_ENCODING))

(defn headers [^HttpRequest req]
  (let [headers (.getHeaders req)
        keys (map (comp s/lower-case key) headers)
        vals (map val headers)]
    (zipmap keys vals)))

(defn create-ring-request [^ChannelHandlerContext context ^HttpRequest http-request]
  (let [[uri query] (url (.getUri http-request))]
    {:body (ChannelBufferInputStream. (.getContent http-request))
     :uri uri
     :query-string query
     :request-method (method (.getMethod http-request))
     :server-name (server-name context http-request)
     :server-port (.getPort (local-address context))
     :remote-addr (remote-address context)
     :scheme (scheme http-request)
     :content-type (content-type http-request)
     :content-length (content-length http-request)
     :character-encoding (character-encoding http-request)
     :headers (headers http-request)}))


src/netty/ring/writers.clj
(ns netty.ring.writers
  (:require [clojure.java.io :as io])
  (:import [org.jboss.netty.buffer ChannelBuffers ChannelBuffer]
           [org.jboss.netty.channel Channel ChannelFutureListener ChannelFuture DefaultFileRegion]
           [org.jboss.netty.handler.codec.http HttpResponse HttpHeaders HttpHeaders$Names]
           [org.jboss.netty.handler.stream ChunkedStream ChunkedFile]
           [java.io InputStream File RandomAccessFile]
           [java.net URLConnection]
           [java.nio.charset Charset]
           [clojure.lang ISeq]))

(def charset (Charset/defaultCharset))

(defn- keep-alive? [^HttpResponse response]
  (= "keep-alive" (HttpHeaders/getHeader response "connection")))

(defn- add-close-listener [^ChannelFuture future ^HttpResponse response]
  (if (keep-alive? response)
    (.addListener future ChannelFutureListener/CLOSE_ON_FAILURE)
    (.addListener future ChannelFutureListener/CLOSE)))

(defn- add-close-stream-listener [^ChannelFuture future ^InputStream stream]
  (let [listener (reify ChannelFutureListener (operationComplete [_ _] (.close stream)))]
    (.addListener future listener)
    (.addListener future ChannelFutureListener/CLOSE)))

(defn- write-response [^HttpResponse response ^Channel channel]
  (-> (.write channel response)
    (add-close-listener response)))

(defprotocol ResponseWriter
  "Provides the best way to write a response for the give ring response body"
  (write [body ^HttpResponse response ^Channel channel]))

(extend-type String
  ResponseWriter
  (write [body ^HttpResponse response ^Channel channel]
    (let [buffer (ChannelBuffers/copiedBuffer body charset)]
      (HttpHeaders/setContentLength response (.readableBytes buffer))
      (.setContent response buffer)
      (write-response response channel))))

(extend-type ISeq
  ResponseWriter
  (write [body ^HttpResponse response ^Channel channel]
    (write (apply str body) response channel)))

(extend-type InputStream
  ResponseWriter
  (write [body ^HttpResponse response ^Channel channel]
    (.write channel response)
    (-> (.write channel (ChunkedStream. body))
      (add-close-stream-listener body))))

(def ^:dynamic *zero-copy* false)

(defn file-body [file]
  (let [random-access-file (RandomAccessFile. file "r")]
    (if *zero-copy*
      (DefaultFileRegion. (.getChannel random-access-file) 0 (.length file) true)
      (ChunkedFile. random-access-file))))

(extend-type File
  ResponseWriter
  (write [body ^HttpResponse response ^Channel channel]
    (let [response-body (file-body body)]
      (.setHeader response "Zero-Copy" *zero-copy*)
      (HttpHeaders/setContentLength response (.length body))

      (.write channel response)
      (add-close-listener (.write channel response-body) response))))

(extend-type nil
  ResponseWriter
  (write [body ^HttpResponse response ^Channel channel]
    (HttpHeaders/setContentLength response 0)
    (write-response response channel)))


src/netty/ring/response.clj
(ns netty.ring.response
  (:require [netty.ring.writers :as w])
  (:import [org.jboss.netty.handler.codec.http
            HttpResponseStatus
            HttpVersion
            HttpHeaders
            HttpResponse
            DefaultHttpResponse]
           [org.jboss.netty.channel
            ChannelHandlerContext
            ChannelFutureListener]))

(defn set-headers [^HttpResponse response headers]
  (doseq [[key values] headers]
    (.setHeader response key values)))

(defn write-ring-response [^ChannelHandlerContext context ring-response]
  (let [status (HttpResponseStatus/valueOf (ring-response :status 200))
        response (DefaultHttpResponse. HttpVersion/HTTP_1_1 status)]
    (set-headers response (:headers ring-response))
    (w/write (:body ring-response) response (.getChannel context))))


src/netty/ring/adapter.clj
(ns netty.ring.adapter
  (:require [netty.ring.request :as request]
            [netty.ring.response :as response]
            [netty.ring.writers :as writers]
            [clojure.tools.logging :as log])
  (:import [java.util.concurrent Executors]
           [java.net InetSocketAddress]
           [org.jboss.netty.channel.socket.nio NioServerSocketChannelFactory]
           [org.jboss.netty.bootstrap ServerBootstrap]
           [org.jboss.netty.handler.stream ChunkedWriteHandler]
           [org.jboss.netty.handler.execution ExecutionHandler OrderedMemoryAwareThreadPoolExecutor]
           [org.jboss.netty.handler.logging LoggingHandler]
           [org.jboss.netty.logging InternalLoggerFactory]
           [org.jboss.netty.channel ChannelHandlerContext Channels SimpleChannelUpstreamHandler ChannelPipelineFactory]
           [org.jboss.netty.handler.codec.http HttpRequestDecoder HttpResponseEncoder HttpChunkAggregator HttpHeaders]))

(def default-options
  {:port 8080
   :zero-copy false
   :channel-options {"child.tcpNoDelay" true "child.keepAlive" true "reuseAddress" true "backlog" 200}
   :max-http-chunk-length 1048576
   :number-of-handler-threads 16
   :max-channel-memory-size 1048576
   :max-total-memory-size 1048576})

(defn- add-keep-alive [http-request ring-response]
  (if (HttpHeaders/isKeepAlive http-request)
    (assoc-in ring-response [:headers "connection"] "keep-alive")
    ring-response))

(defn- create-handler-factory [handler zero-copy]
  #(proxy [SimpleChannelUpstreamHandler] []
     (messageReceived [context event]
       (binding [writers/*zero-copy* zero-copy]
         (let [http-request (.getMessage event)]
           (->> http-request
             (request/create-ring-request context)
             handler
             (add-keep-alive http-request)
             (response/write-ring-response context)))))
     (exceptionCaught [context evt]
       (log/error (.getCause evt) "Error occurred in I/O thread")
       (when (-> context .getChannel .isOpen)
         (response/write-ring-response context {:status 500})))))

(defn- create-logging-factory [debug-type]
  (case debug-type
    :commons (org.jboss.netty.logging.CommonsLoggerFactory.)
    :jboss (org.jboss.netty.logging.JBossLoggerFactory.)
    :log4j (org.jboss.netty.logging.Log4JLoggerFactory.)
    :slf4j (org.jboss.netty.logging.Slf4JLoggerFactory.)
    (org.jboss.netty.logging.JdkLoggerFactory.)))

(defn- pipeline-factory [handler execution-handler options]
  (reify ChannelPipelineFactory
    (getPipeline [this]
      (let [pipeline (Channels/pipeline)]
        (when (:debug options) (.addLast pipeline "logger" (LoggingHandler.)))

        (doto pipeline
          (.addLast "decoder" (HttpRequestDecoder.))
          (.addLast "chunkedAggregator" (HttpChunkAggregator. (:max-http-chunk-length options)))
          (.addLast "encoder" (HttpResponseEncoder.))
          (.addLast "chunkedWriter" (ChunkedWriteHandler.))
          (.addLast "execution" execution-handler)
          (.addLast "handler" (handler)))))))

(defn- create-bootstrap [channel-factory pipeline options]
  (doto (ServerBootstrap. channel-factory)
    (.setPipelineFactory pipeline)
    (.setOptions options)))

(defn- create-execution-handler [options]
  (let [{:keys [number-of-handler-threads max-channel-memory-size max-total-memory-size]} options
        executor (OrderedMemoryAwareThreadPoolExecutor. number-of-handler-threads max-channel-memory-size max-total-memory-size)]
    (ExecutionHandler. executor)))

(defn start-server
  ([handler] (start-server handler {}))
  ([handler user-options]
    (when (:debug user-options)
      (InternalLoggerFactory/setDefaultFactory (create-logging-factory (:debug user-options))))

    (let [options (merge default-options user-options)
          channel-factory (NioServerSocketChannelFactory.)
          handler (create-handler-factory handler (:zero-copy options))
          execution-handler (create-execution-handler options)
          pipeline (pipeline-factory handler execution-handler options)
          bootstrap (create-bootstrap channel-factory pipeline (:channel-options options))
          bind-address (InetSocketAddress. (:port options))
          channel (.bind bootstrap bind-address)]
      (fn []
        (.close channel)
        (.releaseExternalResources channel-factory)
        (.releaseExternalResources execution-handler)))))